File: rust/scopedb-client/src/statement.rs
// Copyright 2024 ScopeDB, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::time::Duration;

use exn::IntoExn;
use exn::Result;
use jiff::SignedDuration;
use uuid::Uuid;

use crate::Error;
use crate::StatementCancelResponse;
use crate::client::Client;
use crate::protocol::Response;
use crate::protocol::ResultFormat;
use crate::protocol::StatementEstimatedProgress;
use crate::protocol::StatementRequest;
use crate::protocol::StatementRequestParams;
use crate::protocol::StatementResponse;
use crate::result::ResultSet;

#[derive(Debug)]
pub struct Statement {
    client: Client,
    statement: String,
    statement_id: Option<Uuid>,
    exec_timeout: Option<SignedDuration>,
    format: ResultFormat,
}

impl Statement {
    pub fn with_statement_id(mut self, statement_id: Uuid) -> Self {
        self.statement_id = Some(statement_id);
        self
    }

    pub fn with_exec_timeout(mut self, exec_timeout: SignedDuration) -> Self {
        self.exec_timeout = Some(exec_timeout);
        self
    }

    pub async fn submit(self) -> Result<StatementHandle, Error> {
        let Statement {
            client,
            statement,
            statement_id,
            exec_timeout,
            format,
        } = self;

        let resp = client
            .submit_statement(StatementRequest {
                statement,
                statement_id,
                exec_timeout,
                params: StatementRequestParams { format },
            })
            .await?;

        match resp {
            Response::Success(response) => Ok(StatementHandle {
                client,
                statement_id: response.statement_id(),
                format,
                response: Some(response),
            }),
            Response::Failed(err) => {
                Err(Error(format!("failed to submit statement: {err}")).into_exn())
            }
        }
    }

    pub(crate) fn new(client: Client, statement: String) -> Self {
        Self {
            client,
            statement,
            statement_id: None,
            exec_timeout: None,
            format: ResultFormat::Json,
        }
    }
}

#[derive(Debug)]
pub struct StatementHandle {
    client: Client,
    statement_id: Uuid,
    format: ResultFormat,
    response: Option<StatementResponse>,
}

impl StatementHandle {
    pub fn statement_id(&self) -> Uuid {
        self.statement_id
    }

    pub fn status(&self) -> Option<&str> {
        self.response.as_ref().map(|r| r.status())
    }

    pub fn is_terminated(&self) -> bool {
        self.response.as_ref().is_some_and(|r| r.is_terminated())
    }

    pub fn progress(&self) -> Option<&StatementEstimatedProgress> {
        self.response.as_ref().map(|r| r.progress())
    }

    pub fn result_set(&self) -> Option<ResultSet> {
        let result_set = self.response.as_ref().and_then(|r| r.result_set())?;
        Some(ResultSet::from_statement_result_set(result_set.clone()))
    }

    pub async fn fetch_once(&mut self) -> Result<(), Error> {
        if self.is_terminated() {
            return Ok(());
        }

        let resp = self
            .client
            .fetch_statement(
                self.statement_id,
                StatementRequestParams {
                    format: self.format,
                },
            )
            .await?;

        match resp {
            Response::Success(response) => {
                self.response = Some(response);
                Ok(())
            }
            Response::Failed(err) => {
                Err(Error(format!("failed to submit statement: {err}")).into_exn())
            }
        }
    }

    pub async fn fetch(mut self) -> Result<ResultSet, Error> {
        while !self.is_terminated() {
            tokio::time::sleep(Duration::from_millis(50)).await;
            self.fetch_once().await?;
        }

        match self.response.unwrap() {
            StatementResponse::Finished { result_set, .. } => {
                Ok(ResultSet::from_statement_result_set(result_set.clone()))
            }
            StatementResponse::Failed { message, .. } => {
                Err(Error(format!("statement failed: {message}")).into_exn())
            }
            StatementResponse::Cancelled { message, .. } => {
                Err(Error(format!("statement cancelled: {message}")).into_exn())
            }
            StatementResponse::Pending { .. } => {
                unreachable!("pending statements should not be fetched")
            }
            StatementResponse::Running { .. } => {
                unreachable!("running statements should not be fetched")
            }
        }
    }

    pub async fn cancel(&mut self) -> Result<StatementCancelResponse, Error> {
        if let Some(response) = self.response.as_ref() {
            match response {
                StatementResponse::Pending { .. } | StatementResponse::Running { .. } => {}
                StatementResponse::Finished {
                    statement_id,
                    created_at,
                    ..
                } => {
                    return Ok(StatementCancelResponse {
                        statement_id: *statement_id,
                        created_at: *created_at,
                        status: "finished".to_string(),
                        message: "statement is finished".to_string(),
                    });
                }
                StatementResponse::Failed {
                    statement_id,
                    created_at,
                    ..
                } => {
                    return Ok(StatementCancelResponse {
                        statement_id: *statement_id,
                        created_at: *created_at,
                        status: "failed".to_string(),
                        message: "statement is failed".to_string(),
                    });
                }
                StatementResponse::Cancelled {
                    statement_id,
                    created_at,
                    ..
                } => {
                    return Ok(StatementCancelResponse {
                        statement_id: *statement_id,
                        created_at: *created_at,
                        status: "cancelled".to_string(),
                        message: "statement is cancelled".to_string(),
                    });
                }
            }
        }

        match self.client.cancel_statement(self.statement_id).await? {
            Response::Success(response) => Ok(response),
            Response::Failed(err) => {
                Err(Error(format!("failed to cancel statement: {err}")).into_exn())
            }
        }
    }

    pub(crate) fn new(client: Client, statement_id: Uuid, format: ResultFormat) -> Self {
        Self {
            client,
            statement_id,
            format,
            response: None,
        }
    }
}

File: node/src/statement.ts
/*
 * Copyright 2024 ScopeDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { Client, RequestOptions } from "./client.js";
import { ScopeDBError } from "./errors.js";
import type {
  ResultFormat,
  StatementCancelResult,
  StatementEstimatedProgress,
  StatementStatus,
} from "./protocol.js";
import { statementIsFinished, statementIsTerminated } from "./protocol.js";
import type { Value } from "./result.js";
import { ResultSet } from "./result.js";

export interface FetchOptions extends RequestOptions {
  /**
   * Initial polling delay in milliseconds.
   * The delay doubles on each poll up to `maxDelayMs`.
   * @default 5
   */
  initialDelayMs?: number;
  /**
   * Maximum polling delay in milliseconds.
   * @default 1000
   */
  maxDelayMs?: number;
}

export class Statement {
  private statementIdValue?: string;
  private execTimeoutValue?: string;
  private maxParallelismValue?: number;
  private readonly format: ResultFormat = "json";

  constructor(
    private readonly client: Client,
    private readonly statementText: string,
  ) {}

  withStatementId(statementId: string): this {
    this.statementIdValue = statementId;
    return this;
  }

  withExecTimeout(execTimeout: string): this {
    this.execTimeoutValue = execTimeout;
    return this;
  }

  withMaxParallelism(maxParallelism: number): this {
    this.maxParallelismValue = maxParallelism;
    return this;
  }

  async submit(options: RequestOptions = {}): Promise<StatementHandle> {
    const status = await this.client.submitStatement(
      {
        statement: this.statementText,
        statement_id: this.statementIdValue,
        exec_timeout: this.execTimeoutValue,
        max_parallelism: this.maxParallelismValue,
        format: this.format,
      },
      options,
    );

    return new StatementHandle(this.client, status.statement_id, this.format, status);
  }

  async execute(options: FetchOptions = {}): Promise<ResultSet> {
    const handle = await this.submit(options);
    return handle.fetch(options);
  }

  /**
   * Executes the statement and returns the first row as a plain object, or
   * `null` if the result set is empty.
   *
   * Useful for point lookups and aggregate queries that return at most one row.
   *
   * @example
   * const row = await client.statement("SELECT count(*) AS n FROM events").executeOne();
   * console.log(row?.["n"]); // bigint
   */
  async executeOne(options: FetchOptions = {}): Promise<Record<string, Value> | null> {
    return (await this.execute(options)).first();
  }
}

export class StatementHandle {
  constructor(
    private readonly client: Client,
    readonly statementId: string,
    private readonly format: ResultFormat = "json",
    private currentStatus?: StatementStatus,
  ) {}

  status(): StatementStatus | undefined {
    return this.currentStatus;
  }

  progress(): StatementEstimatedProgress | undefined {
    return this.currentStatus?.progress;
  }

  resultSet(): ResultSet | null {
    if (this.currentStatus === undefined || !statementIsFinished(this.currentStatus)) {
      return null;
    }
    return ResultSet.fromStatementResultSet(this.currentStatus.result_set);
  }

  async fetchOnce(options: RequestOptions = {}): Promise<void> {
    if (this.currentStatus !== undefined && statementIsTerminated(this.currentStatus)) {
      return;
    }

    this.currentStatus = await this.client.fetchStatement(this.statementId, options);
  }

  async fetch(options: FetchOptions = {}): Promise<ResultSet> {
    let delayMs = options.initialDelayMs ?? 5;
    const maxDelayMs = options.maxDelayMs ?? 1000;

    for (;;) {
      await this.fetchOnce(options);

      if (this.currentStatus === undefined) {
        throw new ScopeDBError("Unexpected", "statement fetch returned no status");
      }

      switch (this.currentStatus.status) {
        case "finished":
          return ResultSet.fromStatementResultSet(this.currentStatus.result_set);
        case "failed":
        case "cancelled":
          throw new ScopeDBError("StatementFailed", this.currentStatus.message);
        case "pending":
        case "running":
          await sleep(delayMs, options.signal);
          if (delayMs < maxDelayMs) {
            delayMs = Math.min(delayMs * 2, maxDelayMs);
          }
          break;
      }
    }
  }

  async cancel(options: RequestOptions = {}): Promise<StatementCancelResult> {
    if (this.currentStatus !== undefined) {
      switch (this.currentStatus.status) {
        case "finished":
          return {
            statement_id: this.currentStatus.statement_id,
            created_at: this.currentStatus.created_at,
            status: "finished",
            message: "statement is finished",
          };
        case "failed":
          return {
            statement_id: this.currentStatus.statement_id,
            created_at: this.currentStatus.created_at,
            status: "failed",
            message: "statement is failed",
          };
        case "cancelled":
          return {
            statement_id: this.currentStatus.statement_id,
            created_at: this.currentStatus.created_at,
            status: "cancelled",
            message: "statement is cancelled",
          };
        case "pending":
        case "running":
          break;
      }
    }

    const result = await this.client.cancelStatement(this.statementId, options);
    if (result.status === "failed") {
      this.currentStatus = {
        status: "failed",
        statement_id: result.statement_id,
        created_at: result.created_at,
        progress: emptyProgress(),
        message: result.message,
      };
    } else if (result.status === "cancelled") {
      this.currentStatus = {
        status: "cancelled",
        statement_id: result.statement_id,
        created_at: result.created_at,
        progress: emptyProgress(),
        message: result.message,
      };
    }
    return result;
  }
}

function emptyProgress(): StatementEstimatedProgress {
  return {
    total_percentage: 0,
    nanos_from_submitted: 0,
    nanos_from_started: 0,
    total_stages: 0,
    total_partitions: 0,
    total_rows: 0,
    total_compressed_bytes: 0,
    total_uncompressed_bytes: 0,
    scanned_stages: 0,
    scanned_partitions: 0,
    scanned_rows: 0,
    scanned_compressed_bytes: 0,
    scanned_uncompressed_bytes: 0,
    skipped_partitions: 0,
    skipped_rows: 0,
    skipped_compressed_bytes: 0,
    skipped_uncompressed_bytes: 0,
  };
}

async function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  if (signal?.aborted) {
    throw signal.reason ?? new Error("aborted");
  }

  await new Promise<void>((resolve, reject) => {
    const timer = setTimeout(resolve, ms);
    const onAbort = () => {
      clearTimeout(timer);
      reject(signal?.reason ?? new Error("aborted"));
    };

    signal?.addEventListener("abort", onAbort, { once: true });
  });
}
